source/common/router/router.cc

```diff
@@ -1373,6 +1416,19 @@ Filter::UpstreamRequest::UpstreamRequest(Filter& parent, Http::ConnectionPool::I
     }
   }
 
+  if (parent_.cluster_->timeoutBudgetStats().has_value()) {
+    const auto start_time = parent_.callbacks_->dispatcher().timeSource().monotonicTime();
+    timeout_budget_stats_ = std::make_unique<Cleanup>([this, start_time]() {
```

Comment:
Instead of using a Cleanup, can you just send the stat when the per-try completes? From a readability perspective this seems a bit convoluted unless there is a good reason to do it.

Reply:
It seemed more convenient to have it fire with the correct value whenever the try either times out or completes successfully without putting the same stat tracking code in a few different places.  This also lets the stat be cancelled if the global timeout hits first instead of recording a value into the per-try histogram when the per-try timeout didn't play into the request's timeout.  @tonya11en and I had a discussion about this offline and here: https://github.com/envoyproxy/envoy/pull/9227#discussion_r359545490

If it's less readable, I think I have to move this to a few different places.  I can try to figure out another way to preserve the behavior discussed above w/r/t the global timeout cancelling it.

If there's a difference of opinion on what should be in the histograms, I would also change this to address readability concerns.

Reply:
IMO this is a bit magical and I would prefer it be explicit, but I don't feel that strongly about it as long as it has good comments.